x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/TaperedNgramTokenFilter.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.wildcard.mapper;
+
+import org.apache.lucene.analysis.TokenFilter;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+
+import java.io.IOException;
+
+/**
+ * Variation on Lucene's NGramTokenFilter that uses smaller (1 character) ngrams for the final few characters in a string. Helps improve
+ * performance of short suffix queries e.g. "*.exe"
+ */
+public final class TaperedNgramTokenFilter extends TokenFilter {
```

Comment:
Do we want to add normalization like keyword?

Reply:
I don't feel the need to have it from the first release, but I think it'd be nice. :+1: FWIW I don't think it has an impact on using a tokenizer vs. a filter.

Reply:
@markharwood any news on making it a tokenizer? I think it would better fit what we are trying to do since we expect it to work on whole strings instead of being tokenized?

Reply:
Do we still need a custom solution here ? The NGramTokenizer should be enough now that we use 3 characters ngram ?